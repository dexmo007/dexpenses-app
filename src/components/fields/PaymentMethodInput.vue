<template>
  <v-select
    class="payment-method-input"
    :value="value"
    v-bind="$attrs"
    @input="$emit('input', $event)"
    :items="paymentMethods"
    menu-props="auto"
    label="Payment method"
    :prepend-icon="icons[value]"
    hide-details
    single-line
  >
    <template v-slot:item="{item}">
      {{$t('paymentMethods.'+item)}}
    </template>
    <template v-slot:selection="{item}">
      {{$t('paymentMethods.'+item)}}
    </template>
  </v-select>
</template>
<script>
import { paymentMethods } from '@dexpenses/core';
import { paymentMethodIcons } from '@/util/receipt';

export default {
  name: 'PaymentMethodInput',
  props: ['value'],
  data() {
    return {
      paymentMethods,
      icons: paymentMethodIcons,
    };
  },
};
</script>
<style>
.payment-method-input .v-input__slot {
  max-width: 10em;
}
</style>
