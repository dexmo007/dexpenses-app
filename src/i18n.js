import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

// function loadLocaleMessages() {
//   const locales = require.context('./i18n', true, /[A-Za-z0-9-_,\s]+\.json$/i);
//   const messages = {};
//   locales.keys().forEach(key => {
//     const matched = key.match(/([A-Za-z0-9-_]+)\./i);
//     if (matched && matched.length > 1) {
//       const locale = matched[1];
//       messages[locale] = locales(key);
//     }
//   });
//   return messages;
// }
const locale = process.env.VUE_APP_I18N_LOCALE || 'en';

const loadedLanguages = [locale];

VueI18n.prototype.setLanguageAsync = async function setLanguageAsync(lang) {
  if (this.locale === lang) {
    return;
  }
  if (!loadedLanguages.includes(lang)) {
    const msgs = await import(/* webpackChunkName: "lang-[request]" */ `@/i18n/${lang}`);
    this.setLocaleMessage(lang, msgs.default);
    loadedLanguages.push(lang);
  }
  this.locale = lang;
  document.querySelector('html').setAttribute('lang', lang);
};
export default new VueI18n({
  locale,
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: {
    /* eslint-disable-next-line import/no-dynamic-require,global-require */ // default language controlled by .env
    [locale]: require(`@/i18n/${locale}.json`),
  },
});
