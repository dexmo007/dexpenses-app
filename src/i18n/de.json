{
  "dashboard": "Dashboard",
  "recurring": "Wiederkehrend",
  "receipts": "Bons",
  "rules": "Regeln",
  "upload": "Hochladen",
  "settings": "Einstellungen",
  "language": "Spache",
  "phoneNumber": "Telefonnummer",
  "homeAddress": "Heimataddresse",
  "account": "Konto",
  "favorites": "Favoriten",
  "nearby": "In der NÃ¤he"
}
