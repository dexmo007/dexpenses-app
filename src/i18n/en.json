{
  "dashboard": "Dashboard",
  "recurring": "Recurring",
  "receipts": "Receipts",
  "rules": "Rules",
  "upload": "Upload",
  "settings": "Settings",
  "language": "Language",
  "phoneNumber": "Phone Number",
  "homeAddress": "Home Address",
  "account": "Account",
  "favorites": "Favorites",
  "nearby": "Nearby",
  "paymentMethods": {
    "DEBIT": "Debit Card",
    "CREDIT": "Credit Card",
    "CASH": "Cash",
    "DKV_CARD": "Fuel Card",
    "PAYPAL": "PayPal",
    "ONLINE": "Online"
  },
  "fields": {
    "phoneNumber": {
      "name": "Phone number",
      "invalidMessage": "Invalid phone number.",
      "countryCallingCodeHint": "It is recommended to supply the country calling code, e.g. +49."
    },
    "cityCode": "City code",
    "name": "Name",
    "classifier": "Classifier",
    "paymentMethod": "Payment method",
    "category": "Category",
    "notes": "Notes",
    "testDataIdentifier": "Test data identifier"
  },
  "validations": {
    "_default": "{_field_} is invalid.",
    "required": "{_field_} is required.",
    "uniqueTestDataRecord": "{_field_} already exists."
  },
  "admin": {
    "testData": {
      "title": "Test data",
      "upload": "Upload",
      "newRecord": "New record",
      "browse": "Browse"
    }
  },
  "actions": {
    "deleteImage": "Delete image",
    "save": "Save"
  }
}
