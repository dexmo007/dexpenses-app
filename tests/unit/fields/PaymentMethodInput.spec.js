import { mount, createLocalVue } from '@vue/test-utils';
import createVuetify from '../vuetify';
import i18n from '@/i18n';
import PaymentMethodInput from '@/components/fields/PaymentMethodInput.vue';

const localVue = createLocalVue();

describe('PaymentMethodInput.vue', () => {
  let vuetify;
  beforeEach(() => {
    vuetify = createVuetify();
  });

  it('should render correctly', () => {
    const wrapper = mount(PaymentMethodInput, {
      i18n,
      localVue,
      vuetify,
      propsData: { eager: true },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly with value', () => {
    const wrapper = mount(PaymentMethodInput, {
      i18n,
      localVue,
      vuetify,
      propsData: {
        value: 'DEBIT',
        eager: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should trigger input event if user selects new item', () => {
    const wrapper = mount(PaymentMethodInput, {
      i18n,
      localVue,
      vuetify,
      propsData: { eager: true },
    });
    wrapper.find('.v-select__slot').trigger('click');
    wrapper.find('.v-list-item').trigger('click'); // click first item: DEBIT

    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toHaveLength(1);
    expect(wrapper.emitted().input[0][0]).toEqual('DEBIT');
  });

  it('should not trigger input event if new value is set programmatically', () => {
    const wrapper = mount(PaymentMethodInput, {
      i18n,
      localVue,
      vuetify,
      propsData: {
        value: 'DEBIT',
        eager: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
    wrapper.setProps({
      value: 'CREDIT',
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted().input).toBeFalsy();
  });

  it('should update v-model correctly', () => {
    const wrapper = mount(
      {
        template: '<PaymentMethodInput v-model="modelValue" eager/>',
        data: () => ({
          modelValue: null,
        }),
        components: {
          PaymentMethodInput,
        },
      },
      { i18n, localVue, vuetify }
    );

    wrapper.find('.v-select__slot').trigger('click');
    wrapper.find('.v-list-item').trigger('click'); // click first item: DEBIT

    expect(wrapper.vm.modelValue).toEqual('DEBIT');
    expect(wrapper.element).toMatchSnapshot();
  });
});
