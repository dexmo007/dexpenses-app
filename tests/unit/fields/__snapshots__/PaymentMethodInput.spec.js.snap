// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentMethodInput.vue should not trigger input event if new value is set programmatically 1`] = `
<div
  class="v-input payment-method-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        credit_card
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-53"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          
    Debit Card
  
          <input
            aria-readonly="true"
            id="input-53"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-53"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-53"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="debit-list-item-59"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Debit Card
  
              </div>
              <div
                aria-labelledby="credit-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Credit Card
  
              </div>
              <div
                aria-labelledby="cash-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Cash
  
              </div>
              <div
                aria-labelledby="dkv_card-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Fuel Card
  
              </div>
              <div
                aria-labelledby="paypal-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    PayPal
  
              </div>
              <div
                aria-labelledby="online-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Online
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaymentMethodInput.vue should not trigger input event if new value is set programmatically 2`] = `
<div
  class="v-input payment-method-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        credit_card
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-53"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          
    Credit Card
  
          <input
            aria-readonly="true"
            id="input-53"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-53"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-53"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="debit-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Debit Card
  
              </div>
              <div
                aria-labelledby="credit-list-item-59"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Credit Card
  
              </div>
              <div
                aria-labelledby="cash-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Cash
  
              </div>
              <div
                aria-labelledby="dkv_card-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Fuel Card
  
              </div>
              <div
                aria-labelledby="paypal-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    PayPal
  
              </div>
              <div
                aria-labelledby="online-list-item-59"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Online
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaymentMethodInput.vue should render correctly 1`] = `
<div
  class="v-input payment-method-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-4"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <label
          class="v-label theme--light"
          for="input-4"
          style="left: 0px; position: absolute;"
        >
          Payment method
        </label>
        <div
          class="v-select__selections"
        >
          <input
            aria-readonly="true"
            id="input-4"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-4"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-4"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="debit-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Debit Card
  
              </div>
              <div
                aria-labelledby="credit-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Credit Card
  
              </div>
              <div
                aria-labelledby="cash-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Cash
  
              </div>
              <div
                aria-labelledby="dkv_card-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Fuel Card
  
              </div>
              <div
                aria-labelledby="paypal-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    PayPal
  
              </div>
              <div
                aria-labelledby="online-list-item-9"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Online
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaymentMethodInput.vue should render correctly with value 1`] = `
<div
  class="v-input payment-method-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        credit_card
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-20"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          
    Debit Card
  
          <input
            aria-readonly="true"
            id="input-20"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-20"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-20"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="debit-list-item-26"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Debit Card
  
              </div>
              <div
                aria-labelledby="credit-list-item-26"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Credit Card
  
              </div>
              <div
                aria-labelledby="cash-list-item-26"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Cash
  
              </div>
              <div
                aria-labelledby="dkv_card-list-item-26"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Fuel Card
  
              </div>
              <div
                aria-labelledby="paypal-list-item-26"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    PayPal
  
              </div>
              <div
                aria-labelledby="online-list-item-26"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Online
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaymentMethodInput.vue should update v-model correctly 1`] = `
<div
  class="v-input payment-method-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field v-text-field--single-line v-text-field--is-booted v-select primary--text"
  eager=""
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light primary--text"
      >
        credit_card
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-71"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          
    Debit Card
  
          <input
            aria-readonly="true"
            id="input-71"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light primary--text"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 8; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-71"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-71"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="debit-list-item-89"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Debit Card
  
              </div>
              <div
                aria-labelledby="credit-list-item-89"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Credit Card
  
              </div>
              <div
                aria-labelledby="cash-list-item-89"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Cash
  
              </div>
              <div
                aria-labelledby="dkv_card-list-item-89"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Fuel Card
  
              </div>
              <div
                aria-labelledby="paypal-list-item-89"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    PayPal
  
              </div>
              <div
                aria-labelledby="online-list-item-89"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                
    Online
  
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
